### PR TITLE
Add emebedding classifier to rasa nlu

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- add count vectors featurizer
- add embedding intent classifier

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by amn41** [comment]
please also: 

- [ ] add these components to the [registry](https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/registry.py)
- [ ] create a standard pipeline "tensorflow_embedding" with components ` ["tokenizer_whitespace","count_vectors_featurizer", "classifier_tensorflow_embedding"]`

**Comment by Ghostvv** [comment]
there is a problem with `is_training` we need to create `tf.placeholder` due to the new save scheme

**Comment by pawni** [comment]
When changing that I didn't realise that we rely on this - we could also just build the graph as normal and only use the saver to reload the weights?

**Comment by Ghostvv** [comment]
it's ok, we'll load the graph as well, i will just add placeholder with default value during training as true and false during prediction

**Comment by Ghostvv** [comment]
I think we do not need white space tokenizer as sklearn's CountVectorizer do it for us

**Comment by amn41** [comment]
ok! then we can ditch the whitespace tokenizer

**Comment by Ghostvv** [comment]
@pawni could you please take a look at the registry.py file if i did it correct. Also could you please make the correct implementation of default config

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+
+        with self.graph.as_default():
+            a_in = tf.placeholder(tf.float32, (None, X.shape[-1]), name='a')
+            b_in = tf.placeholder(tf.float32, (None, None, Y.shape[-1]),
+                                  name='b')
+
+            self.embedding_placeholder = a_in
+            self.intent_placeholder = b_in
+
+            is_training = tf.placeholder(tf.bool)
```
Comment (pawni) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: I believe building this as `tf.placeholder_with_default` might make it easier - the default could be false and therefore we wouldn't need to save it for processing messages

Comment (Ghostvv) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: I'll change that

File: rasa_nlu/featurizers/count_vectors_featurizer.py [hunk 2]
```diff
@@ -18,85 +18,102 @@
...
-
     def process(self, message, **kwargs):
         # type: (Message, **Any) -> None
-        
+
         bag = self.vect.transform([self._lemmetize(message)]).toarray()
         message.set("text_features", bag)
 
-                                    
     def _lemmetize(self, message):
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa_nlu/featurizers/count_vectors_featurizer.py]: sorry about that

**Comment by Ghostvv** [comment]
it looks like `tensorflow_embedding` pipeline is automatically tested

**Comment by amn41** [comment]
that's good. I think worth adding a test for the `CountVectorFeaturizer` - basically just passing it a few different strings and checking the corresponding vectors you get out makes sense.

e.g.:
`["hello hello hello hello hello "] -> [5]`
`["hello goodbye hello"] -> [2, 1]`
`["a b c d e f"] -> [1,1,1,1,1]`

with the appropriate params (e.g min_occurrence)

File: docs/pipeline.rst [hunk 3]
```diff
@@ -167,6 +193,78 @@ intent_classifier_sklearn
...
+                    "confidence": 0.28161531595656784,
+                    "name": "restaurant_search"
+                }
+            ]
+        }
+
+:Description:
+    The embedding intent classifier embeds user inputs and intent labels into the same space. Supervised embeddings are
+    trained by maximizing similarity between them. This algorithm is based on
+    the starspace idea from: https://arxiv.org/abs/1709.03856.
```
Comment (amn41) [inline docs/pipeline.rst]: mention that by default we have some hidden layers and dropout so it's not quite starspace

File: docs/pipeline.rst [hunk 4]
```diff
@@ -167,6 +193,78 @@ intent_classifier_sklearn
...
+    It also provides rankings of the labels that did not "win". The embedding intent classifier
+    needs to be preceded by a featurizer in the pipeline. This featurizer creates the features used for the embeddings.
```
Comment (amn41) [inline docs/pipeline.rst]: does this work with any featurizer? or just the countvectorfeaturizer?

File: docs/pipeline.rst [hunk 5]
```diff
@@ -104,6 +107,29 @@ intent_featurizer_ngrams
...
+    .. code-block:: yaml
+
+        pipeline:
+        - name: "intent_featurizer_count_vectors"
+          # regular expression for tokens
+          "token_pattern": r'(?u)\b\w\w+\b'
+          # min number of word occurancies in the document to add to vocabulary
+          "min_df": 1
+          # max number (fraction if float) of word occurancies in the document to add to vocabulary
+          "max_df": 1.0
```
Comment (amn41) [inline docs/pipeline.rst]: I think this comment could be clearer about what max_df is , maybe also link to sklearn docs

File: rasa_nlu/featurizers/count_vectors_featurizer.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    requires = []
+
+    defaults = {
+        # regular expression for tokens
+        "token_pattern": r'(?u)\b\w\w+\b',
+
+        # min number of word occurancies in the document to add to vocabulary
+        "min_df": 1,
+
+        # max number (fraction if float) of word occurancies
```
Comment (amn41) [inline rasa_nlu/featurizers/count_vectors_featurizer.py]: spelling: "occurances" (also in comment above)

File: docs/pipeline.rst [hunk 7]
```diff
@@ -167,6 +193,78 @@ intent_classifier_sklearn
...
+    There are several hyperparameters such as neural network's number of hidden layers, embedding dimension,
+    droprate, regularization, etc.
+    In the config, you can specify these parameters.
+
+    .. note:: There is a parameter that controls similarity `similarity_type`.
+              It should be either `cosine` or `inner`. For `cosine` similarity `mu_pos` and `mu_neg`
+              should be between `-1` and `1`. Parameter `mu_pos` controls how similar the algorithm
+              should try to make embedding vectors for correct intent labels,
+              while `mu_neg` controls maximum negative similarity for incorrect intents.
+              It is set to negative value to mimic the original
```
Comment (akelad) [inline docs/pipeline.rst]: "It is set to **a** negative value"

File: docs/pipeline.rst [hunk 8]
```diff
@@ -167,6 +193,78 @@ intent_classifier_sklearn
...
+
+:Description:
+    The embedding intent classifier embeds user inputs and intent labels into the same space. Supervised embeddings are
+    trained by maximizing similarity between them. This algorithm is based on
+    the starspace idea from: https://arxiv.org/abs/1709.03856.
+    It also provides rankings of the labels that did not "win". The embedding intent classifier
+    needs to be preceded by a featurizer in the pipeline. This featurizer creates the features used for the embeddings.
+
+:Configuration:
+    There are several hyperparameters such as neural network's number of hidden layers, embedding dimension,
```
Comment (akelad) [inline docs/pipeline.rst]: "the neural networks"

File: docs/pipeline.rst [hunk 9]
```diff
@@ -167,6 +193,78 @@ intent_classifier_sklearn
...
+
+    .. note:: There is a parameter that controls similarity `similarity_type`.
+              It should be either `cosine` or `inner`. For `cosine` similarity `mu_pos` and `mu_neg`
+              should be between `-1` and `1`. Parameter `mu_pos` controls how similar the algorithm
+              should try to make embedding vectors for correct intent labels,
+              while `mu_neg` controls maximum negative similarity for incorrect intents.
+              It is set to negative value to mimic the original
+              starspace algorithm in the case `mu_neg = mu_pos` and `use_max_sim_neg = False`.
+              See starspace paper https://arxiv.org/abs/1709.03856 for details.
+              If `use_max_sim_neg = True` the algorithm only minimize maximum
```
Comment (akelad) [inline docs/pipeline.rst]: minimizes

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+        "C2": 0.002,
+        "C_emb": 0.8,
+        "droprate": 0.2,
+
+        # flag if to tokenize intents
+        "intent_tokenization_flag": False,
+        "intent_split_symbol": '_'
+    }
+
+    def __init__(self, component_config=None,
```
Comment (akelad) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: this method is suuuuper long. I'd put stuff like loading parameters from config and checking if hidden layer size is valid in separate functions and then call them here.

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+        self.embedding_placeholder = embedding_placeholder
+        self.intent_placeholder = intent_placeholder
+        self.similarity_op = similarity_op
+
+    @classmethod
+    def required_packages(cls):
+        # type: () -> List[Text]
+        return ["tensorflow"]
+
+    def train(self, training_data, cfg=None, **kwargs):
```
Comment (akelad) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: again, a lot of the stuff in this function can go in separate functions to be called from here. As you can probably see in code climate, the maximum function length is technically 25

File: docs/pipeline.rst (hunk 4 above)
Comment (Ghostvv) [reply docs/pipeline.rst]: the following configuration worked:
```
pipeline:
- name: "nlp_spacy"
- name: "tokenizer_spacy"
- name: "intent_featurizer_spacy"
- name: "intent_featurizer_ngrams"
- name: "intent_classifier_tensorflow_embedding"
```
though accuracy was quite low

File: rasa_nlu/classifiers/embedding_intent_classifier.py (hunk 11 above)
Comment (Ghostvv) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: do you think it makes sense to add additional helper functions in load and persist?

**Comment by akelad** [comment]
Ok definitely looks a lot better! @amn41 I think we're good to merge
